File: app/src/main/java/com/itba/runningMate/mainpage/fragments/pastruns/ui/PastRunsPresenter.java
package com.itba.runningMate.mainpage.fragments.pastruns.ui;

import com.itba.runningMate.domain.Run;
import com.itba.runningMate.repository.run.RunRepository;
import com.itba.runningMate.utils.schedulers.SchedulerProvider;

import java.lang.ref.WeakReference;
import java.util.LinkedList;
import java.util.List;

import io.reactivex.disposables.Disposable;
import timber.log.Timber;

public class PastRunsPresenter {

    private final WeakReference<PastRunsView> view;
    private final RunRepository runRepository;
    private final SchedulerProvider schedulerProvider;

    private Disposable disposable;


    public PastRunsPresenter(final SchedulerProvider schedulerProvider, final RunRepository runRepository, PastRunsView view) {
        this.view = new WeakReference<>(view);
        this.runRepository = runRepository;
        this.schedulerProvider = schedulerProvider;
    }

    public void onViewAttached() {
        disposable = runRepository.getRunLazy()
                .subscribeOn(schedulerProvider.computation())
                .observeOn(schedulerProvider.ui())
                .subscribe(this::receivedRunList, this::onRunListError);

    }

    private void onRunListError(final Throwable err) {
        Timber.d("Failed to retrieve runs from db");
        if (view.get() != null) {
            view.get().showNoPastRunsMessage();
        }
    }

    public void onViewDetached() {
        disposable.dispose();
    }

    public void receivedRunList(List<Run> runs) {
        if (view.get() == null) {
            return;
        }
        if (runs == null || runs.isEmpty()) {
            view.get().showNoPastRunsMessage();
            view.get().updatePastRuns(new LinkedList<>());
        } else {
            view.get().hideNoPastRunsMessage();
            view.get().updatePastRuns(runs);
        }
    }

    public void refreshAction() {
        this.onViewAttached();
    }

    public void onRunClick(long id) {
        if (view.get() != null) {
            view.get().launchRunDetails(id);
        }
    }
}

File: app/src/main/java/com/itba/runningMate/pastruns/PastRunsPresenter.kt
package com.itba.runningMate.pastruns

import com.itba.runningMate.domain.Run
import com.itba.runningMate.repository.aggregaterunmetrics.AggregateRunMetricsStorage
import com.itba.runningMate.repository.run.RunRepository
import com.itba.runningMate.utils.providers.schedulers.SchedulerProvider
import io.reactivex.disposables.CompositeDisposable
import timber.log.Timber
import java.lang.ref.WeakReference
import java.util.*

class PastRunsPresenter(
    private val schedulerProvider: SchedulerProvider,
    private val runRepository: RunRepository,
    private val aggregateRunMetricsStorage: AggregateRunMetricsStorage,
    view: PastRunsView
) {
    private val view: WeakReference<PastRunsView> = WeakReference(view)
    private val disposables: CompositeDisposable = CompositeDisposable()

    fun onViewAttached() {
        fetchRuns()
    }

    fun onViewDetached() {
        disposables.dispose()
    }

    fun fetchRuns() {
        disposables.add(runRepository.getRunLazy()
            .subscribeOn(schedulerProvider.computation())
            .observeOn(schedulerProvider.ui())
            .subscribe({ runs: List<Run>? -> receivedRunList(runs) }) { onRunListError() }
        )
    }

    private fun receivedRunList(runs: List<Run>?) {
        if (view.get() == null) {
            return
        }
        if (runs == null || runs.isEmpty()) {
            view.get()!!.showNoPastRunsMessage()
            view.get()!!.updatePastRuns(LinkedList())
        } else {
            view.get()!!.hideNoPastRunsMessage()
            view.get()!!.updatePastRuns(runs)
        }
    }

    private fun onRunListError() {
        Timber.d("Failed to retrieve runs from db")
        if (view.get() != null) {
            view.get()!!.showNoPastRunsMessage()
        }
    }

    fun onRunClick(id: Long) {
        if (view.get() != null) {
            view.get()!!.launchRunDetails(id)
        }
    }

    fun onSwipeRunToDelete(id: Long) {
        disposables.add(runRepository.deleteRun(id)
            .subscribeOn(schedulerProvider.computation())
            .observeOn(schedulerProvider.ui())
            .subscribe({ onRunDeleted(id) }) { onRunDeleteError(id) })
    }

    private fun onRunDeleted(id: Long) {
        Timber.i("Successfully deleted run from db for run-id: %d", id)
    }

    private fun onRunDeleteError(id: Long) {
        Timber.d("Failed to delete run from db for run-id: %d", id)
        if (view.get() != null) {
            view.get()!!.showDeleteError()
        }
    }
}